doc/arch/014-reporting-etl-pipeline.md

```diff
@@ -0,0 +1,150 @@
+# Reporting ETL Pipeline
+
+## Context
+
+Simple's dashboard reports are currently generated by queries on the application database's tables. This has several
+disadvantages as we start to scale.
+
+### 1. Complexity
+
+Simple's application schema is optimized for application transactions like registering patients, recording BPs,
+updating patient info, or tracking overdue patients. As such, reporting queries grow complex. Asking questions like
+"latest BP in a month", "prescribed drugs as of a date", "current step of treatment protocol" involve complex clauses
+in the SQL queries.
+
+This complexity has a tangible impact on the quality of our reports. We risk data inconsistency as complex queries are
+more error-prone when reused in multiple places.
+
+
+### 2. Coupling
+
+When the application schema is used for reporting also, there is a tight coupling between application requirements and
+reporting requirements. Changes to the schema to optimize for reporting may have a negative impact on complexity in the
+application, and vice versa.
+
+In addition to complexity, performance is also a growing concern as we scale. Heavy reporting operations may strain the
+application's ability to write data to the database. Conversely, heavy application activity may impact our ability to
+generate reports in a performant manner. We currently use heavy layers of caching to mitigate these performance
+concerns, but the less caching we _need_ to do, the better.
```

Comment:
> but the less caching we _need_ to do, the better.

I think it's worth elaborating on _why_ since this is our current approach.

Caching inherently comes with the added complexity of cache invalidation, which is touted to be one of the harder problems in computer science. Here are some reasons I can think of:
- We're past the point of using simple caching, and we're layering complex abstractions on it.
- We have a few use cases like drug stock, merging duplicate patients, etc where caching is not always an option.
- Our data can be changed retroactively (new data in the past, and updates to measurements), so cache invalidation can potentially become complex.